### ISSUE TITLE
Potential bug in br_if validation

### ISSUE THREAD

**Issue by kripken** [WebAssembly/wabt#588]
I thought this was a bug in SpiderMonkey as wabt and v8 accept the binary, but @lukewagner says it is the opposite. Details here: https://bugzilla.mozilla.org/show_bug.cgi?id=1385717

Seems like it comes down to what is pushed on the stack by a br_if with a value whose input value is unreachable, but this is a confusing part of the spec at least for me so I'm not sure what is right. Looks like v8 and wabt do one thing with sm and the spec interpreter do the other.

**Comment by binji** [comment]
Yeah, this is a bug in wabt/v8, similar to the one with `tee_local`. Luke's explanation on the bug is correct.

**Comment by eholk** [comment]
I agree with Luke's explanation as well.

**Comment by kripken** [comment]
To make sure I understand: the `br_if` pops an unreachable as the value it will send. It then pushes something else - an `i32` - which is then invalid to leave on the stack there. So it is pushing something different than what it popped, I presume because it knows it is branching to a block with type `i32`?

**Comment by binji** [comment]
Here's how I think of it: `br_if` always pops an `i32` and `t` (which is determined by the label's signature). In this example `t` is `i32`. The unreachable doesn't change this, it just means that there is potentially more than one valid concrete stack. The code is valid if and only if there is at least one concrete stack that could pass validation. It doesn't really make sense to talk about popping unreachable, IMO. You're always popping a concrete type. So when we validate the code here we're trying to figure out the constraints of a valid concrete stack. 

Here's a simplified example of the same bug:
```
block (result i32)
  block
    unreachable
    br_if 1
  end
end
```

You can think of the `unreachable` as having a signature where it can consume anything on the stack, and produce anything else. That's what Luke meant by `[t*]->[t*]`. So let's just say that after the unreachable instruction, the stack looks like `[...]`.

`br_if` has the signature `[t i32]->[t]`, but as mentioned above, we know that `t` here is `i32`. Therefore we know that for `br_if` to be valid, the stack must have `[i32 i32]` on the top. So we can put constraints on the stack after the `unreachable` instruction -- it has to look like `[... i32 i32]`. After the `br_if`, the condition value is consumed, so the stack looks like `[... i32]`. It doesn't really matter whether the other value was popped and repushed, just that it has this structure.

Finally, the block is closed. We know that for the block to be valid there can't be anything on the stack. But we know that it looks like `[... i32]`. So this snippet can't be valid -- there's no concrete stack where this instruction sequence would properly type-check.

**Comment by jgravelle-google** [comment]
To try and explain (and to make sure *I* understand):

tl;dr the `br_if` needs an `i32`, the top of the stack has `unreachable` which creates the `i32`, and if the `br_if` isn't taken it doesn't pop the created `i32` meaning it should fail validation.

Even shorter: what Luke said yeah.

Long version:

Given the wasm function from @binji's PR:
```
block (result i32)
  block
    unreachable
    br_if 1
  end
  i32.const 1
end
```
Then we can trace the value stack through the decoding as follows:

value stack after instr | wasm instr | instr signature
--|--|--
[] | `block (result i32)` | [] -> [`i32`]
[] | `block` | [] -> []
[`t*`] | `unreachable` | [] -> []
[`t*`, `i32`, `i32`] | (before `br_if`, synthesize two i32s) | -
[`t*`, `i32`] | `br_if 1` | [`i32`, `i32`] -> [`i32`]
[`t*`, `i32`] (**validation failure**) | (before `end`, expecting either empty or `t*`) | -
[] | `end` (void block) | [] -> []
[`i32`] | `i32.const 1` | [] -> [`i32`]
[] | `end` (`i32` block) | [`i32`] -> []

Where `t*` is the `unreachable` type.

**Comment by binji** [comment]
FYI: filed v8 bug here: https://bugs.chromium.org/p/v8/issues/detail?id=6651

**Comment by kripken** [comment]
thanks @binji and @jgravelle-google, makes a lot of sense to me now.
